crates/http/src/wagi/util.rs

```diff
@@ -301,17 +308,17 @@ pub fn compose_response(stdout: &[u8]) -> Result<Response<Body>, Error> {
     Ok(res)
 }
 
-fn parse_cgi_headers(headers: String) -> HashMap<String, String> {
-    let mut map = HashMap::new();
+fn parse_cgi_headers(headers: String) -> Vec<(String, String)> {
```

Comment:
I am not sure this change is needed.  The logic in the iterator seems (wisely) order-independent:

* If status comes out (of the Vec or HashMap) _before_ location, location sees that `explicit_status_code` is set, and respects it.
* If status comes out _after_ location, then status overrides the 302 that was set by location.
* If only one of them comes out then we don't have a problem.

Given that, I _think_ you can revert this and the `as_str` changes.  But I'm quite prepared to believe I'm missing something!

Reply:
i am slightly confused with this statement in Spec:

```
The order in which header fields with differing field names are received is not significant. However, it is "good practice" to send general-header fields first, followed by request-header or response- header fields, and ending with the entity-header fields.

Multiple message-header fields with the same field-name MAY be present in a message if and only if the entire field-value for that header field is defined as a comma-separated list [i.e., #(values)]. It MUST be possible to combine the multiple header fields into one "field-name: field-value" pair, without changing the semantics of the message, by appending each subsequent field-value to the first, each separated by a comma. The order in which header fields with the same field-name are received is therefore significant to the interpretation of the combined field value, and thus a proxy MUST NOT change the order of these field values when a message is forwarded.
```

Although I should also confess that I implemented this change as I thought headers order was required for the status code (because `golang` wanted me to call `WriteHeader(code)` before adding any other header). 

I can make this change tomorrow, but in the meantime, if anyone has the cycles to take this ticket over, I will be happy to close this PR and let other PR go through.

Reply:
Because we are working like a proxy between the response sent by the `wagi` module and the client, we should potentially preserve the order of headers when sending the response back?

Reply:
cc @lann @radu-matei what do you think?

Reply:
The CGI spec certainly implies that ordering matters, at least for the first header(s): https://datatracker.ietf.org/doc/html/rfc3875#section-6.2

```
CGI-Response = document-response | local-redir-response |
               client-redir-response | client-redirdoc-response

document-response = Content-Type [ Status ] *other-field NL
                    response-body

client-redir-response = client-Location *extension-field NL

client-redirdoc-response = client-Location Status Content-Type
                           *other-field NL response-body
```

For `client-redirdoc-response`:
> The Status header field MUST be supplied and MUST contain a status
> value of 302 'Found', or it MAY contain an extension-code, that is,
> another valid status code that means client redirection.

It seems ambiguous whether `document-response` is allowed to have a `Location` header, though certainly there are protocols out there that expect to send a `Location` with a `2xx` response.